### PR TITLE
Fix an incorrect assertion when the block locator is at the tip

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra panics when it's at the tip, and it's asked for a block locator for that tip.
(This might depend on whether there's a stop height, and whether integer checks are optimised out.)

This might have been triggered by receiving block hash gossips from the new Zebra code in PR #2729.

## Solution

- Skip the assertion when the response is empty
- Add tests for `BlockLocator`, `FindBlockHashes`, and `FindBlockHeaders`
- Add a specific test for this bug

## Review

Anyone can review this panic fix, but I think @oxarbitrage might have looked at this code recently?

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [x] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
The macOS failure looks like a spurious timing issue:
> Message:  Opening database "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/zebrad_tests.qiCrlWIEdGK7/state/state/v10/mainnet" failed: Error { message: "IO error: While lock file: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/zebrad_tests.qiCrlWIEdGK7/state/state/v10/mainnet/LOCK: Resource temporarily unavailable" }. Hint: Check if another zebrad process is running. Try changing the state cache_dir in the Zebra config.

https://github.com/ZcashFoundation/zebra/pull/2789/checks?check_run_id=3692472245#step:10:1326

File: zebra-state/src/service.rs [hunk 1]
```diff
@@ -411,7 +412,8 @@ impl StateService {
     ///   * adding the `stop` hash to the list, if it is in the best chain, or
     ///   * adding `max_len` hashes to the list.
     ///
-    /// Returns an empty list if the state is empty.
+    /// Returns an empty list if the state is empty,
+    /// or the `intersection` is the best chain tip.
```
Comment (oxarbitrage) [inline zebra-state/src/service.rs]:
```suggestion
    /// or if the `intersection` is the best chain tip.
```

File: zebra-state/src/service.rs [hunk 2]
```diff
@@ -489,11 +501,12 @@ impl StateService {
                 .unwrap_or(true),
             "the list must not contain the intersection hash"
         );
-        assert!(
-            stop.map(|hash| !res[..(res.len() - 1)].contains(&hash))
-                .unwrap_or(true),
-            "if the stop hash is in the list, it must be the final hash"
-        );
+        if let (Some(stop), Some((_, res_except_last))) = (stop, res.split_last()) {
```
Comment (oxarbitrage) [inline zebra-state/src/service.rs:504]: nice one.

**Review by oxarbitrage — DISMISSED** [review]
Looks good. Just a minor suggestion in a doc comment.

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**